mempool/src/pool.rs

```diff
@@ -0,0 +1,1093 @@
+        let inputs = tx
+            .inputs()
+            .iter()
+            .map(|input| {
+                let outpoint = input.outpoint();
+                self.chain_state
+                    .get_outpoint_value(outpoint)
+                    .or_else(|_| self.store.get_unconfirmed_outpoint_value(outpoint))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let sum_inputs = inputs
+            .iter()
+            .cloned()
+            .sum::<Option<_>>()
+            .ok_or(TxValidationError::InputValuesOverflow)?;
+        let sum_outputs = tx
+            .outputs()
+            .iter()
+            .map(|output| output.value())
+            .sum::<Option<_>>()
+            .ok_or(TxValidationError::OutputValuesOverflow)?;
+        (sum_inputs - sum_outputs).ok_or(TxValidationError::InputsBelowOutputs)
+    }
+}
+
+fn get_relay_fee(tx: &Transaction) -> Amount {
+    // TODO we should never reach the expect, but should this be an error anyway?
+    eprintln!("get_relay_fee - encoded_size {:?}", tx.encoded_size());
+    Amount::from_atoms(u128::try_from(tx.encoded_size() * RELAY_FEE_PER_BYTE).expect("Overflow"))
+}
+
+pub trait Mempool<C, T, M> {
+    fn create(chain_state: C, clock: T, memory_usage_estimator: M) -> Self;
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), Error>;
+    fn get_all(&self) -> Vec<&Transaction>;
+    fn contains_transaction(&self, tx: &Id<Transaction>) -> bool;
+    fn drop_transaction(&mut self, tx: &Id<Transaction>);
+    fn new_tip_set(&mut self, chain_state: C);
+}
+
+pub trait ChainState: Debug + 'static {
+    fn contains_outpoint(&self, outpoint: &OutPoint) -> bool;
+    fn get_outpoint_value(&self, outpoint: &OutPoint) -> Result<Amount, anyhow::Error>;
+}
+
+trait TryGetFee {
+    fn try_get_fee(&self, tx: &Transaction) -> Result<Amount, TxValidationError>;
+}
+
+newtype! {
+    #[derive(Debug)]
+    struct Ancestors(BTreeSet<H256>);
+}
+
+newtype! {
+    #[derive(Debug)]
+    struct Descendants(BTreeSet<H256>);
+}
+
+newtype! {
+    #[derive(Debug)]
+    struct Conflicts(BTreeSet<H256>);
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct TxMempoolEntry {
+    tx: Transaction,
+    fee: Amount,
+    parents: BTreeSet<H256>,
+    children: BTreeSet<H256>,
+    count_with_descendants: usize,
+    fees_with_descendants: Amount,
+    creation_time: Time,
+}
+
+impl TxMempoolEntry {
+    fn new(
+        tx: Transaction,
+        fee: Amount,
+        parents: BTreeSet<H256>,
+        creation_time: Time,
+    ) -> TxMempoolEntry {
+        Self {
+            tx,
+            fee,
+            parents,
+            children: BTreeSet::default(),
+            count_with_descendants: 1,
+            creation_time,
+            fees_with_descendants: fee,
+        }
+    }
+
+    fn count_with_descendants(&self) -> usize {
+        self.count_with_descendants
+    }
+
+    fn tx_id(&self) -> H256 {
```

Comment:
Originally `H256` was used since it implements `Hash` and can be used as the key for a hashmap. However, I eventually wound up using mostly `BTreeSets`, so actually we can use `Id<Transaction>`. Made this change now.